src/System.Text.Json/tests/JsonObjectTests.cs

```diff
@@ -0,0 +1,621 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Text.Json.Tests
+{
+    public static partial class JsonObjectTests
+    {
+        [Fact]
+        public static void TestDefaultConstructor()
+        {
+            var jsonObject = new JsonObject();
+            Assert.Equal(0, jsonObject.PropertyNames.Count);
+            Assert.Equal(0, jsonObject.PropertyValues.Count);
+        }
+
+        [Fact]
+        public static void TestIEnumerableKVPConstructor()
+        {
+            var jsonProperties = new List<KeyValuePair<string, JsonNode>>();
+            jsonProperties.Add(new KeyValuePair<string, JsonNode>("number", new JsonNumber(17)));
+            jsonProperties.Add(new KeyValuePair<string, JsonNode>("string", new JsonString("property value")));
+            jsonProperties.Add(new KeyValuePair<string, JsonNode>("boolean", new JsonBoolean(true)));
+
+            var jsonObject = new JsonObject(jsonProperties);
+
+            Assert.IsType<JsonNumber>(jsonObject["number"]);
+            Assert.Equal(17, (jsonObject["number"] as JsonNumber).GetInt32());
```

Comment:
Is there really value in the IsType test?  If the key was not present that's throwing an exception, and I'm not sure what it does for null... but the Assert.Equal really seems to do it all.  If JsonNumber was not `sealed` then it would be showing that we returned a particular level in the type hierarchy... but I'm pretty sure it's sealed...

```C#
// Doesn't throw, returns right type, and has the right value:
Assert.Equal(17, ((JsonNumber)jsonObject["number"]).GetInt32());
```

Reply:
(This was initially a "why are we reading from the dictionary twice?" question, then I simplified)